external-crates/move/crates/move-binary-format/src/compatibility.rs

```diff
@@ -360,98 +362,165 @@ pub enum InclusionCheck {
-                .zip(&new_enum.variants)
-                .all(|(old, new)| old == new)
-            {
-                return err;
-            }
-        }
+        // function checks
 
-        // Function checks
-        for (name, old_func) in &old_module.functions {
-            match new_module
-                .functions
-                .get(name)
-                .or_else(|| new_module.functions.get(name))
-            {
-                Some(new_func) if old_func == new_func => (),
-                _ => {
-                    return err;
+        compare_identifiers(old_module.functions.keys(), new_module.functions.keys()).for_each(
+            |mark| match mark {
+                Mark::New(name) => {
+                    context.function_new(name, new_module.functions.get(name).unwrap())
+                }
+                Mark::Missing(name) => context.function_missing(name),
+                Mark::Existing(name) => {
+                    let old_function = old_module.functions.get(name).unwrap();
+                    let new_function = new_module.functions.get(name).unwrap();
+                    if old_function != new_function {
+                        context.function_change(name, old_function);
+                    }
+                }
+            },
+        );
+
+        // use iters for friends
+        let mut old_friend_iter = old_module.friends.iter().peekable();
+        let mut new_friend_iter = new_module.friends.iter().peekable();
+        loop {
+            let old_friend = old_friend_iter.peek();
+            let new_friend = new_friend_iter.peek();
+            match (old_friend, new_friend) {
+                (Some(old_friend), Some(new_friend)) => {
+                    if old_friend != new_friend {
+                        if old_friend.lt(new_friend) {
+                            context.friend_missing(old_friend);
+                            old_friend_iter.next();
+                        } else {
+                            context.friend_new(new_friend);
+                            new_friend_iter.next();
+                        }
+                    } else {
+                        old_friend_iter.next();
+                        new_friend_iter.next();
+                    }
+                }
+                (Some(old_friend), None) => {
+                    context.friend_missing(old_friend);
+                    old_friend_iter.next();
+                }
+                (None, Some(new_friend)) => {
+                    context.friend_new(new_friend);
+                    new_friend_iter.next();
+                }
+                (None, None) => {
+                    break;
                 }
             }
         }
 
-        Ok(())
+        context.finish(self)
     }
 }
+
+enum Mark<'a> {
+    New(&'a Identifier),
+    Missing(&'a Identifier),
+    Existing(&'a Identifier),
+}
+
+fn compare_identifiers<'c, I, J>(old: I, new: J) -> impl Iterator<Item = Mark<'c>> + 'c
```

Comment:
This needs lots of unit tests.

Reply:
going to try proptest, seems worth it for this area.